sdk/core/azure-core/inc/azure/core/context.hpp

```diff
@@ -349,21 +146,26 @@ namespace Azure { namespace Core {
      * @return A value associated with the context found; an empty value if a specific value can't
      * be found.
      */
-    const ContextValue& operator[](const std::string& key) const
+    template <class T> const T& Get(const std::string& key) const
```

Comment:
Replace string key with Context::Key const&, which is a class with empty constructor, and internally we would look up on the address of contextKey.
You declare ContextKeys mostly in .cpp files, where they are `static const`. If it needs to be shared, they are declared in headers as `AZ_*_DLLEXPORT extern Azure::Context::Key;`
Internally, we would store `&contextKey`, contract being that if `&lhs == &rhs`, they are equal, otherwise they are not?

Reply:
An idea, `Context::Key` can even have a typedef inside indicating values of which type are stored under it. That would effectively limit us to one type per key, but Get<T>() should become safer.

Reply:
What is `Context::Key`? is it a typedef of `std::string`?

Reply:
@Jinming-Hu that would be a nested class, it can be defined as simple as `class Key {};` inside `Context` class (so that we don't pollute Azure namespace with the classes that are useless for customer intellisense, and can be used only on connection with `Context` and not much else). Typedef string woud not be as good, because it means that our comparisons during lookup need to be string comparisons, and I propose pointer comparisons.

Reply:
But I think pointer comparisons will be problematic. Two keys can have same values but different addresses, right?

Reply:
Keys would be defined as
```
namespace
{
   Azure::Context::Key RetryKey;
}

void RetryPolicy::InitializeContextOptions(Context& context, RetryOptions const& options)
{
   context[RetryKey] = options;
}

void RetryPolicy::Send(Context const& context, ...)
{
   auto options = context.GetOrDefault(RetryKey, m_options);
   ...
}
```

Or, if within a header:
```
namespace Azure { namespace Core { Namespace Http { Namespace Policy {
{
   AZ_CORE_DLLEXPORT extern Azure::Context::Key const RetryContextKey;

   // Now, RetryContextKey is available to customers. Just for example:
   inline void RetryPolicy::InitializeContextOptions(Context& context, RetryOptions const& options)
   {
      context[RetryKey] = options;
   }
   ...
```
+ in .cpp file:
```
Azure::Context::Key const Azure::Core::Http::Policy::RetryContextKey;

void RetryPolicy::Send(Context const& context, ...)
{
   auto options = context.GetOrDefault(RetryKey, m_options);
   ...
}
```
potentially, in customer code:
```
int main()
{
    Context ctx;
    ctx[Azure::Core::Http::RetryContextKey] = RetryOptions(...);
}
```

Class `Context::Key` would work like this:
context.hpp
```
class Context
{
public:
   class Key;

   operator [] (Key const&)
   T const& GetOrDefault(Key const&, T const& default);
}
```

context.cpp
```
bool operator == (Context::Key const& lhs, Context::Key const& rhs)
{
   return &lhs == &rhs;
}
```

Reply:
This looks like kind of complicated. I don't see the advantage. I still think `std::string` as key is just fine

Reply:
I am leaving this up to Rick, this is why I am not blocking, but the benefit would be way faster comparisons (pointers vs strings), and preventing from collisions/injections (if Key is privately defined inside .cpp, there is no way to use it outside). Jeffrey was for the variation of this idea too (he was proposing `void*`s; we basically do the same but with type safety).

Reply:
> but the benefit would be way faster comparisons (pointers vs strings)

This is definitely a premature optimization, we already have tons of string copy, string comparison. Plus, as @JeffreyRichter said, we should not worry about string copy in a network program.

> he was proposing void*s; we basically do the same but with type safety

I remember that was for value, not key

Reply:
We did have a review recently, he proposed for Key as well.

Reply:
> We did have a review recently, he proposed for Key as well.

I'm so sad to hear that 😂.

Reply:
> This is definitely a premature optimization

@Jinming-Hu  I have seen this phrase often around the PRs, what's our criteria for using this classification ?  I wonder if we might be over-using the term and loosing opportunities to make the code to perform better even on simple things to do.